File: app/macos/hyperwhisper/Managers/Transcription/Support/RustHTTPExecutor.swift
//
//  RustHTTPExecutor.swift
//  hyperwhisper
//
//  Shared sans-I/O HTTP plumbing for the Rust shared core (Wave 3 / M3-B).
//
//  The Rust core builds a fully-described `HttpRequest` value (URL, method,
//  headers, body) and parses an `HttpResponse` value; the PLATFORM owns all
//  actual network I/O. Audio bytes NEVER cross the FFI boundary — a file is
//  referenced by path (`Body.fileStream` / `HwPart.fileRef` /
//  `Body.jsonWithBase64File`) and the platform streams it from disk. That holds
//  even for the inline-base64 body: the core names the path and the surrounding
//  JSON, and the platform does the encoding.
//
//  This executor takes a binding `HttpRequest`, performs the I/O with
//  `URLSession`, and returns a binding `HttpResponse`. It is reused by EVERY
//  cloud STT provider, so the five `Body` cases below must be exactly right —
//  every later provider inherits any bug here.
//
//  `HttpRequest`, `HttpResponse`, `Header`, `HttpMethod`, `Body`, `HwPart` are
//  all UniFFI-generated types in the `HyperWhisper` module (see
//  `RustCore/hyperwhisper_core.swift`).
//

import Foundation

enum RustHTTPExecutor {

    /// Sentinel multipart field name marking a single-`fileRef` body that must be
    /// streamed as the **raw** request body (not a `multipart/form-data`
    /// envelope). MUST stay byte-identical to `RAW_BODY_FIELD` in
    /// `shared-core-rs/crates/hw-net/src/providers/hyperwhisper_cloud.rs`:
    ///
    ///     pub const RAW_BODY_FIELD: &str = "@raw";
    ///
    /// HyperWhisper Cloud + the routed (Azure-MAI / Google-Chirp) providers
    /// encode their raw-streamed upload as `Body.multipart` carrying exactly one
    /// `HwPart.fileRef(field: "@raw", …)`. We detect that shape and `upload`
    /// the file as the bare request body, with `Content-Type = fileRef.mime`.
    static let rawBodyField = "@raw"

    /// Perform `request` over `session` and capture an `HttpResponse` for the
    /// core to parse.
    ///
    /// - Throws `CancellationError` if the Swift `Task` is cancelled before the
    ///   request is issued. URLSession network errors propagate untranslated so
    ///   the retry wrapper can classify them.
    static func execute(_ request: HttpRequest, session: URLSession) async throws -> HttpResponse {
        if Task.isCancelled { throw CancellationError() }

        var urlRequest = try buildURLRequest(from: request)

        let data: Data
        let response: URLResponse

        switch request.body {
        case .empty:
            (data, response) = try await session.data(for: urlRequest)

        case let .bytes(_, payload):
            // Content-Type for `.bytes` is applied in `buildURLRequest` (below).
            urlRequest.httpBody = payload
            (data, response) = try await session.data(for: urlRequest)

        case let .fileStream(path, _):
            // Raw file body. Content-Type is applied in `buildURLRequest`.
            // `upload(for:fromFile:)` streams from disk — audio is never buffered
            // into memory across FFI.
            let fileURL = URL(fileURLWithPath: path)
            (data, response) = try await session.upload(for: urlRequest, fromFile: fileURL)

        case let .multipart(boundary, parts):
            if let rawFile = rawStreamFileRef(in: parts) {
                // === @raw SENTINEL PATH (HW Cloud / routed) ===
                // A single fileRef whose field == "@raw" means: stream this file
                // as the RAW request body, NOT a multipart envelope. The
                // Content-Type is the fileRef's own mime (set in buildURLRequest
                // when it detected the @raw shape).
                let fileURL = URL(fileURLWithPath: rawFile.path)
                (data, response) = try await session.upload(for: urlRequest, fromFile: fileURL)
            } else {
                // === REAL multipart/form-data PATH (used by the next sub-module:
                // OpenAI / Groq / ElevenLabs / Mistral / Grok) ===
                // Assemble the envelope from the core-provided parts, in order,
                // using the core-provided boundary. File parts are streamed from
                // disk via a temp envelope so audio bytes still never cross FFI.
                //
                // Reserve the temp URL and register cleanup BEFORE writing, so a
                // mid-write throw (e.g. a missing fileRef.path) does not leak the
                // partial temp file.
                let bodyFileURL = FileManager.default.temporaryDirectory
                    .appendingPathComponent("hw-multipart-\(UUID().uuidString).tmp")
                defer { try? FileManager.default.removeItem(at: bodyFileURL) }
                try writeMultipartBody(to: bodyFileURL, boundary: boundary, parts: parts)
                urlRequest.setValue(
                    "multipart/form-data; boundary=\(boundary)",
                    forHTTPHeaderField: "Content-Type"
                )
                (data, response) = try await session.upload(for: urlRequest, fromFile: bodyFileURL)
            }

        case let .jsonWithBase64File(prefix, path, suffix):
            // === INLINE-BASE64 JSON PATH (Gemini 3.5 Transcribe) ===
            // `/v1beta/interactions` has no file-reference form, so the core
            // hands us the two JSON fragments that surround the audio and we
            // splice the base64 in between. Same temp-file discipline as the
            // multipart path: reserve the URL and register cleanup BEFORE
            // writing, and encode in chunks so a 14 MB file is never held twice
            // in memory. Content-Type is applied in `buildURLRequest`.
            let bodyFileURL = FileManager.default.temporaryDirectory
                .appendingPathComponent("hw-jsonb64-\(UUID().uuidString).tmp")
            defer { try? FileManager.default.removeItem(at: bodyFileURL) }
            try writeJSONWithBase64Body(to: bodyFileURL, prefix: prefix, path: path, suffix: suffix)
            (data, response) = try await session.upload(for: urlRequest, fromFile: bodyFileURL)
        }

        guard let http = response as? HTTPURLResponse else {
            throw TranscriptionError.invalidResponse(details: "Non-HTTP response")
        }

        return HttpResponse(
            status: UInt16(clamping: http.statusCode),
            headers: responseHeaders(from: http),
            body: data
        )
    }

    // MARK: - Request construction

    /// Build the base `URLRequest`: URL, method, all core headers, and the
    /// `Content-Type` for body shapes that carry it inline (`.bytes`,
    /// `.fileStream`, and the `@raw` multipart). For real multipart the
    /// `Content-Type` (with boundary) is set at upload time.
    /// Materialise a core request into a `URLRequest` with its body attached.
    ///
    /// Use this instead of `execute` when the caller owns the I/O: the LLM
    /// post-processing paths each set their own timeout, run their own retry
    /// policy, and one of them reads an SSE stream line by line. They still get
    /// the URL, the headers and the body from the core (issue #282).
    ///
    /// Only inline bodies are attached. A file-backed body must go through
    /// `execute`, which streams it from disk — and asking for one here THROWS
    /// rather than returning a bodyless request. Dropping it silently would put
    /// a fully-formed, correctly-authenticated request with an empty payload on
    /// the wire, which reads at the call site as a provider rejecting valid
    /// input. Unreachable today (every caller builds a JSON `.bytes` body), so
    /// the switch is exhaustive with no `default` arm: a new `Body` case has to
    /// be classified here rather than falling into the silent path.
    static func buildInlineURLRequest(from request: HttpRequest) throws -> URLRequest {
        var urlRequest = try buildURLRequest(from: request)
        switch request.body {
        case .empty:
            break
        case let .bytes(_, payload):
            urlRequest.httpBody = payload
        case .fileStream, .multipart, .jsonWithBase64File:
            throw TranscriptionError.serverError(
                statusCode: 0,
                message: "buildInlineURLRequest cannot attach a file-backed body — route this request through RustHTTPExecutor.execute"
            )
        }
        return urlRequest
    }

    static func buildURLRequest(from request: HttpRequest) throws -> URLRequest {
        guard let url = URL(string: request.url) else {
            throw TranscriptionError.serverError(statusCode: 0, message: "Invalid request URL: \(request.url)")
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = methodString(request.method)

        // Apply every core-provided header verbatim, in order.
        for header in request.headers {
            urlRequest.setValue(header.value, forHTTPHeaderField: header.name)
        }

        // Body-derived Content-Type for the inline-typed shapes. The core also
        // emits an explicit `Content-Type` header for the @raw multipart (the
        // audio mime), so the loop above already covers that; we additionally
        // guarantee the bytes/fileStream cases set it even if a builder relied on
        // the body's content_type rather than a header.
        switch request.body {
        case let .bytes(contentType, _):
            urlRequest.setValue(contentType, forHTTPHeaderField: "Content-Type")
        case let .fileStream(_, contentType):
            urlRequest.setValue(contentType, forHTTPHeaderField: "Content-Type")
        case let .multipart(_, parts):
            if let rawFile = rawStreamFileRef(in: parts) {
                urlRequest.setValue(rawFile.mime, forHTTPHeaderField: "Content-Type")
            }
            // real multipart: Content-Type set at upload time (needs boundary).
        case .jsonWithBase64File:
            // The spliced body is a JSON document by construction. The core also
            // emits an explicit `Content-Type` header, so this is belt-and-braces
            // for the same reason the bytes/fileStream arms are.
            urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        case .empty:
            break
        }

        return urlRequest
    }

    /// Detect the `@raw` sentinel: a multipart body with exactly ONE part that is
    /// a `fileRef` whose field name equals `rawBodyField`. Returns the file's
    /// `(path, mime, filename)` when matched, else `nil`.
    ///
    /// This is the load-bearing branch the verifier must confirm — getting it
    /// wrong wraps the audio in a multipart envelope and the backend 400s /
    /// transcribes garbage.
    private static func rawStreamFileRef(in parts: [HwPart]) -> (path: String, mime: String, filename: String)? {
        guard parts.count == 1, case let .fileRef(field, path, mime, filename) = parts[0],
              field == rawBodyField else {
            return nil
        }
        return (path: path, mime: mime, filename: filename)
    }

    /// Map the binding `HttpMethod` to the URLRequest method string.
    private static func methodString(_ method: HttpMethod) -> String {
        switch method {
        case .get: return "GET"
        case .post: return "POST"
        case .put: return "PUT"
        case .delete: return "DELETE"
        }
    }

    /// Flatten an `HTTPURLResponse`'s header fields into the binding `Header`
    /// list the core expects. The core's `HttpResponse.header(_:)` lookup is
    /// case-insensitive, so casing here is irrelevant.
    private static func responseHeaders(from http: HTTPURLResponse) -> [Header] {
        http.allHeaderFields.compactMap { key, value in
            guard let name = key as? String else { return nil }
            return Header(name: name, value: String(describing: value))
        }
    }

    // MARK: - Real multipart assembly (next sub-module's providers)

    /// Assemble a `multipart/form-data` body into `tempURL`, streaming each
    /// `fileRef` part from disk so audio bytes never live fully in memory across
    /// FFI. Parts are written in the core-provided order using the core-provided
    /// boundary. The caller owns `tempURL`'s lifecycle (creates the cleanup
    /// `defer` before calling), so a throw mid-write never leaks the partial file.
    private static func writeMultipartBody(to tempURL: URL, boundary: String, parts: [HwPart]) throws {
        FileManager.default.createFile(atPath: tempURL.path, contents: nil)

        let handle = try FileHandle(forWritingTo: tempURL)
        defer { try? handle.close() }

        let crlf = "\r\n"
        func write(_ string: String) throws {
            guard let data = string.data(using: .utf8) else { return }
            try handle.write(contentsOf: data)
        }

        for part in parts {
            switch part {
            case let .field(name, value):
                try write("--\(boundary)\(crlf)")
                try write("Content-Disposition: form-data; name=\"\(name)\"\(crlf)\(crlf)")
                try write("\(value)\(crlf)")

            case let .fileRef(field, path, mime, filename):
                try write("--\(boundary)\(crlf)")
                try write("Content-Disposition: form-data; name=\"\(field)\"; filename=\"\(filename)\"\(crlf)")
                try write("Content-Type: \(mime)\(crlf)\(crlf)")
                // Stream the file in chunks rather than loading it whole.
                let fileHandle = try FileHandle(forReadingFrom: URL(fileURLWithPath: path))
                defer { try? fileHandle.close() }
                while true {
                    let chunk = try fileHandle.read(upToCount: 1 << 20) ?? Data()
                    if chunk.isEmpty { break }
                    try handle.write(contentsOf: chunk)
                }
                try write(crlf)
            }
        }

        try write("--\(boundary)--\(crlf)")
    }

    // MARK: - Inline-base64 JSON assembly (Gemini 3.5 Transcribe)

    /// Read size for the chunked base64 encoder. 3-byte-aligned so a full chunk
    /// encodes without padding; 1.5 MB in → 2 MB of base64 out. Not `private`
    /// so the tests can straddle the boundary the carry logic turns on.
    static let base64ChunkSize = 3 * 512 * 1024

    /// Write `prefix` ++ base64(file at `path`) ++ `suffix` into `tempURL`.
    ///
    /// The encoding is the STANDARD alphabet, padded, with no line breaks —
    /// `Data.base64EncodedData()`'s default — which is what the core's
    /// `Body.JsonWithBase64File` contract specifies. Anything else (URL-safe
    /// alphabet, wrapped lines, missing padding) makes Google reject the audio.
    ///
    /// The file is encoded in chunks so a 14 MB recording is never resident
    /// twice. Padding may only appear at the very END of a base64 stream, so
    /// each chunk written out must be a multiple of 3 raw bytes; the remainder
    /// is carried over to the next read and the final tail (which may be 1 or 2
    /// bytes) is encoded last, where its padding is correct. A short read from
    /// `FileHandle` therefore cannot corrupt the encoding.
    ///
    /// The caller owns `tempURL`'s lifecycle (creates the cleanup `defer` before
    /// calling), matching `writeMultipartBody`.
    ///
    /// Not `private`, so `RustHTTPExecutorBase64Tests` can exercise the carry
    /// across the chunk boundary. A carry bug corrupts only recordings larger
    /// than `chunkSize` (1.5 MB), which no other test reaches.
    static func writeJSONWithBase64Body(
        to tempURL: URL,
        prefix: Data,
        path: String,
        suffix: Data
    ) throws {
        FileManager.default.createFile(atPath: tempURL.path, contents: nil)

        let handle = try FileHandle(forWritingTo: tempURL)
        defer { try? handle.close() }

        try handle.write(contentsOf: prefix)

        let fileHandle = try FileHandle(forReadingFrom: URL(fileURLWithPath: path))
        defer { try? fileHandle.close() }

        let chunkSize = base64ChunkSize
        var pending = Data()
        while true {
            let chunk = try fileHandle.read(upToCount: chunkSize) ?? Data()
            if chunk.isEmpty { break }
            pending.append(chunk)

            let encodable = pending.count - (pending.count % 3)
            if encodable > 0 {
                try handle.write(contentsOf: Data(pending.prefix(encodable)).base64EncodedData())
                pending.removeFirst(encodable)
            }
        }
        // Final 1–2 byte tail: the only place padding is allowed.
        if !pending.isEmpty {
            try handle.write(contentsOf: pending.base64EncodedData())
        }

        try handle.write(contentsOf: suffix)
    }
}

File: app/windows/HyperWhisper/Services/Transcription/RustHttpExecutor.cs
// RUST HTTP EXECUTOR (Wave 3 / Win-2)
//
// Shared sans-I/O HTTP plumbing for the Rust shared core. The Rust core builds a
// fully-described `HttpRequest` value (URL, method, headers, body) and parses an
// `HttpResponse` value; the PLATFORM owns all actual network I/O. Audio bytes
// NEVER cross the FFI boundary — a file is referenced by path
// (`Body.FileStream` / `HwPart.FileRef`) and the platform streams it from disk.
//
// This executor takes a binding `HttpRequest`, performs the I/O with `HttpClient`
// (+ `StreamContent`/`FileStream`), and returns a binding `HttpResponse`. It is
// reused by EVERY cloud STT provider, so the five `Body` cases below must be
// exactly right — every later provider inherits any bug here. It mirrors the
// already-shipped macOS `RustHTTPExecutor.swift` 1:1.
//
// The one exception to "audio never crosses the boundary" is
// `Body.JsonWithBase64File` (Gemini 3.5 Transcribe): that vendor demands the
// audio inline in the JSON, so the PLATFORM reads and base64-encodes it here.
// Rust still only ever sees the path.
//
// TODO-verify (Windows/CI): Rust shared-core swap — compile-only on macOS; not
// built against the C# binding here. Verify under `dotnet build` in CI.

using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using HyperWhisper.Models;

// Binding types (`HttpRequest`, `HttpResponse`, `Header`, `Body`, `HwPart`,
// `HttpMethod`) live in this namespace. They are `internal`, which is fine — the
// app and the binding compile into a single assembly. We qualify aggressively
// below because several names collide with `System.Net.Http.*` and
// `HyperWhisper.*` types.
using uniffi.hyperwhisper_core;

// Disambiguate the binding `HttpMethod` from `System.Net.Http.HttpMethod` for the
// whole file. The binding type is the one the core emits.
using RustHttpMethod = uniffi.hyperwhisper_core.HttpMethod;

namespace HyperWhisper.Services.Transcription;

/// <summary>
/// Executes a Rust-core-built <see cref="HttpRequest"/> over an
/// <see cref="HttpClient"/> and captures an <see cref="HttpResponse"/> for the
/// core to parse. Reused by all 12 cloud STT providers.
/// </summary>
internal static class RustHttpExecutor
{
    /// <summary>
    /// Sentinel multipart field name marking a single-<c>FileRef</c> body that
    /// must be streamed as the RAW request body (not a <c>multipart/form-data</c>
    /// envelope). MUST stay byte-identical to <c>RAW_BODY_FIELD</c> in
    /// <c>shared-core-rs/crates/hw-net/src/providers/hyperwhisper_cloud.rs</c>:
    /// <code>pub const RAW_BODY_FIELD: &amp;str = "@raw";</code>
    ///
    /// HyperWhisper Cloud + the routed (Azure-MAI / Google-Chirp) providers encode
    /// their raw-streamed upload as <c>Body.Multipart</c> carrying exactly one
    /// <c>HwPart.FileRef(field: "@raw", …)</c>. We detect that shape and stream the
    /// file as the bare request body, with <c>Content-Type = fileRef.mime</c>.
    /// </summary>
    internal const string RawBodyField = "@raw";

    /// <summary>
    /// Perform <paramref name="request"/> over <paramref name="client"/> and
    /// capture an <see cref="HttpResponse"/> for the core to parse.
    /// </summary>
    /// <remarks>
    /// Throws <see cref="OperationCanceledException"/> when the token is cancelled.
    /// Transport errors (<see cref="HttpRequestException"/>, timeout
    /// <see cref="TaskCanceledException"/>) propagate untranslated so the retry
    /// wrapper can classify them.
    /// </remarks>
    internal static async Task<HttpResponse> ExecuteAsync(
        HttpRequest request,
        HttpClient client,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        // Build the message + content together so the content (and any underlying
        // FileStream — multipart file parts, @raw raw-stream, .FileStream) is
        // disposed with the request after SendAsync completes. No temp envelope is
        // written to disk: MultipartFormDataContent streams each file part live.
        using var message = BuildRequestMessage(request);

        using var response = await client.SendAsync(
            message,
            HttpCompletionOption.ResponseContentRead,
            cancellationToken).ConfigureAwait(false);

        var body = await response.Content
            .ReadAsByteArrayAsync(cancellationToken)
            .ConfigureAwait(false);

        return new HttpResponse(
            @status: (ushort)(int)response.StatusCode,
            @headers: CaptureHeaders(response),
            @body: body);
    }

    // MARK: - Request construction

    /// <summary>
    /// Build the <see cref="HttpRequestMessage"/>: URL, method, all core headers,
    /// and the body (dispatched by the five <see cref="Body"/> cases).
    /// </summary>
    /// <remarks>
    /// internal (not private): test seam for HyperWhisper.SmokeTests via
    /// InternalsVisibleTo (see HyperWhisper.csproj). The inline-base64 body is
    /// produced lazily as it is written to the socket, so the only way to prove
    /// the bytes are what the vendor expects is to build the message and read its
    /// content back.
    /// </remarks>
    internal static HttpRequestMessage BuildRequestMessage(HttpRequest request)
    {
        var message = new HttpRequestMessage(MapMethod(request.@method), request.@url);

        // Attach the body content FIRST so per-content Content-Type from the core
        // headers below can override (or coexist with) the body's content type.
        switch (request.@body)
        {
            case Body.Empty:
                // No content.
                break;

            case Body.Bytes bytes:
            {
                var content = new ByteArrayContent(bytes.@data);
                TrySetContentType(content, bytes.@contentType);
                message.Content = content;
                break;
            }

            case Body.FileStream fileStream:
            {
                // Raw file body streamed from disk — audio is never buffered into
                // memory across FFI. The StreamContent owns the FileStream; both
                // are disposed when `message` is disposed after SendAsync.
                var stream = OpenFileForStreaming(fileStream.@path);
                var content = new StreamContent(stream);
                TrySetContentType(content, fileStream.@contentType);
                message.Content = content;
                break;
            }

            case Body.JsonWithBase64File json:
            {
                // === INLINE BASE64 JSON PATH (Gemini 3.5 Transcribe) ===
                // The vendor's /v1beta/interactions endpoint has no file-reference
                // form, so the audio must sit inside the JSON. The core hands us
                // two literal JSON fragments and a path; the body is
                // prefix ++ base64(file) ++ suffix.
                //
                // STREAMED, not buffered. The obvious version — ReadAllBytes, then
                // ToBase64String, then a concatenated array — holds the file, its
                // base64, and the joined result live at once: measured 88.7 MB of
                // large-object heap for a 14 MB recording, which is the provider's
                // own cap, and paid again on every RustRetry attempt (up to 8).
                // JsonWithBase64FileContent encodes as it writes to the socket
                // instead, matching the shared-.NET twin (RustHttpTransport) and
                // macOS's chunked RustHTTPExecutor. The file read is async there
                // and runs under SendAsync, so it also honours the attempt's
                // cancellation token, which File.ReadAllBytes on this thread did
                // not.
                //
                // C# `switch` over the binding's Body records is NOT exhaustive:
                // omitting this case would send a body-less POST and the vendor
                // would 400. There is no compiler check here — see the default arm.
                var content = new JsonWithBase64FileContent(json.@prefix, json.@path, json.@suffix);
                TrySetContentType(content, "application/json");
                message.Content = content;
                break;
            }

            case Body.Multipart multipart:
            {
                var rawFile = RawStreamFileRef(multipart.@parts);
                if (rawFile != null)
                {
                    // === @raw SENTINEL PATH (HW Cloud / routed) ===
                    // A single FileRef whose field == "@raw" means: stream this
                    // file as the RAW request body, NOT a multipart envelope. The
                    // Content-Type is the fileRef's own mime.
                    var stream = OpenFileForStreaming(rawFile.@path);
                    var content = new StreamContent(stream);
                    TrySetContentType(content, rawFile.@mime);
                    message.Content = content;
                }
                else
                {
                    // === REAL multipart/form-data PATH ===
                    // (OpenAI / Groq / ElevenLabs / Mistral) Assemble the envelope
                    // from the core-provided parts, in order, using the
                    // core-provided boundary. File parts are streamed from disk so
                    // audio bytes still never cross FFI.
                    message.Content = BuildMultipartContent(multipart.@boundary, multipart.@parts);
                }
                break;
            }

            default:
                // This switch IS the request body. C# switches over the binding's
                // Body records are not exhaustive, so a variant added to the core
                // and not handled here would fall straight through and send a
                // BODY-LESS request — a bare 400 from the vendor with nothing
                // naming the cause. This repo has been bitten by that shape
                // before. Fail loudly instead; the shared-.NET twin
                // (HyperWhisper.SharedCore/RustHttpTransport) has the same arm.
                throw new NotSupportedException(
                    $"Unhandled Rust request body variant: {request.@body.GetType().Name}");
        }

        // Apply every core-provided header verbatim, in order. Headers that belong
        // on the content (Content-Type) are routed to the content; everything else
        // goes on the request. This matches the macOS executor, which sets every
        // header on the URLRequest (URLSession routes content headers internally).
        foreach (var header in request.@headers)
        {
            ApplyHeader(message, header.@name, header.@value);
        }

        return message;
    }

    /// <summary>
    /// Detect the <c>@raw</c> sentinel: a multipart body with exactly ONE part
    /// that is a <c>FileRef</c> whose field name equals <see cref="RawBodyField"/>.
    /// Returns the <c>FileRef</c> when matched, else null.
    ///
    /// This is the load-bearing branch the verifier must confirm — getting it
    /// wrong wraps the audio in a multipart envelope and the backend 400s /
    /// transcribes garbage.
    /// </summary>
    private static HwPart.FileRef? RawStreamFileRef(List<HwPart> parts)
    {
        if (parts.Count == 1 && parts[0] is HwPart.FileRef fileRef && fileRef.@field == RawBodyField)
        {
            return fileRef;
        }
        return null;
    }

    /// <summary>
    /// Assemble a streamed <c>multipart/form-data</c> body. File parts are streamed
    /// from disk via <see cref="StreamContent"/>; field parts are inline strings.
    /// Parts are added in the core-provided order with the core-provided boundary.
    /// </summary>
    private static MultipartFormDataContent BuildMultipartContent(string boundary, List<HwPart> parts)
    {
        var content = new MultipartFormDataContent(boundary);
        foreach (var part in parts)
        {
            switch (part)
            {
                case HwPart.Field field:
                    content.Add(new StringContent(field.@value), field.@name);
                    break;

                case HwPart.FileRef fileRef:
                {
                    var stream = OpenFileForStreaming(fileRef.@path);
                    var fileContent = new StreamContent(stream);
                    if (MediaTypeHeaderValue.TryParse(fileRef.@mime, out var mediaType))
                    {
                        fileContent.Headers.ContentType = mediaType;
                    }
                    content.Add(fileContent, fileRef.@field, fileRef.@filename);
                    break;
                }
            }
        }
        return content;
    }

    /// <summary>
    /// A <see cref="Body.JsonWithBase64File"/> body written as
    /// <c>prefix ++ base64(file) ++ suffix</c>, encoded straight onto the request
    /// stream instead of assembled in memory.
    ///
    /// The prefix/suffix are already valid JSON fragments emitted by the core;
    /// base64 is ASCII by construction, so a byte-level splice cannot corrupt
    /// either side's escaping. <see cref="ToBase64Transform"/> emits the standard
    /// padded, non-URL-safe alphabet with no line breaks — exactly what
    /// <see cref="Convert.ToBase64String"/> produced before, and what the core's
    /// fragments assume.
    ///
    /// Mirrors <c>JsonWithBase64FileContent</c> in
    /// <c>app/shared-dotnet/HyperWhisper.SharedCore/RustHttpTransport.cs</c>.
    /// </summary>
    private sealed class JsonWithBase64FileContent : HttpContent
    {
        private readonly byte[] _prefix;
        private readonly string _path;
        private readonly byte[] _suffix;

        internal JsonWithBase64FileContent(byte[] prefix, string path, byte[] suffix)
        {
            _prefix = prefix;
            _path = path;
            _suffix = suffix;
        }

        // HttpClient calls the cancellable overload, so the token that reaches
        // here is the one ExecuteAsync was given — which is how the file read
        // finally honours a cancelled retry attempt.
        protected override Task SerializeToStreamAsync(Stream stream, TransportContext? context) =>
            WriteAsync(stream, CancellationToken.None);

        protected override Task SerializeToStreamAsync(
            Stream stream,
            TransportContext? context,
            CancellationToken cancellationToken) => WriteAsync(stream, cancellationToken);

        private async Task WriteAsync(Stream stream, CancellationToken cancellationToken)
        {
            await stream.WriteAsync(_prefix, cancellationToken).ConfigureAwait(false);

            await using (var file = OpenFileForStreaming(_path))
            {
                var encoder = new CryptoStream(stream, new ToBase64Transform(), CryptoStreamMode.Write, leaveOpen: true);
                await using (encoder.ConfigureAwait(false))
                {
                    await file.CopyToAsync(encoder, cancellationToken).ConfigureAwait(false);
                    // Emits the padding for a trailing partial group. Explicit
                    // rather than left to Dispose so it provably happens before
                    // the suffix is written.
                    await encoder.FlushFinalBlockAsync(cancellationToken).ConfigureAwait(false);
                }
            }

            await stream.WriteAsync(_suffix, cancellationToken).ConfigureAwait(false);
        }

        protected override bool TryComputeLength(out long length)
        {
            length = 0;
            long fileBytes;
            try
            {
                fileBytes = new FileInfo(_path).Length;
            }
            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
            {
                // Fall back to chunked transfer rather than failing here; the send
                // itself will surface a real error if the file is gone.
                return false;
            }

            // Base64 is 4 output bytes per 3 input bytes, padded up.
            length = _prefix.LongLength + ((fileBytes + 2) / 3) * 4 + _suffix.LongLength;
            return true;
        }
    }

    /// <summary>
    /// Open an audio file for async streaming uploads. Read-shared so a concurrent
    /// cleanup reader/size-probe doesn't fight the upload.
    /// </summary>
    private static FileStream OpenFileForStreaming(string path)
    {
        return new FileStream(
            path,
            FileMode.Open,
            FileAccess.Read,
            FileShare.Read,
            bufferSize: 81920,
            useAsync: true);
    }

    /// <summary>
    /// Apply a core-provided header to the request, routing content headers
    /// (Content-Type) to the content object as required by
    /// <see cref="HttpRequestMessage"/>.
    /// </summary>
    private static void ApplyHeader(HttpRequestMessage message, string name, string value)
    {
        // Content-Type must live on the content. The core may emit it as a header
        // (e.g. the @raw multipart's audio mime); route it to the content.
        if (string.Equals(name, "Content-Type", StringComparison.OrdinalIgnoreCase))
        {
            if (message.Content != null && MediaTypeHeaderValue.TryParse(value, out var mediaType))
            {
                message.Content.Headers.ContentType = mediaType;
            }
            return;
        }

        // Try request headers first; fall back to content headers (e.g.
        // Content-Length, Content-Disposition) when the framework rejects them as
        // content-only. TryAddWithoutValidation is permissive on header names the
        // backend defines (X-STT-*, xi-api-key, Authorization).
        if (!message.Headers.TryAddWithoutValidation(name, value))
        {
            message.Content?.Headers.TryAddWithoutValidation(name, value);
        }
    }

    private static void TrySetContentType(HttpContent content, string contentType)
    {
        if (MediaTypeHeaderValue.TryParse(contentType, out var mediaType))
        {
            content.Headers.ContentType = mediaType;
        }
    }

    /// <summary>Map the binding <see cref="RustHttpMethod"/> to a framework verb.</summary>
    private static System.Net.Http.HttpMethod MapMethod(RustHttpMethod method) => method switch
    {
        RustHttpMethod.Get => System.Net.Http.HttpMethod.Get,
        RustHttpMethod.Post => System.Net.Http.HttpMethod.Post,
        RustHttpMethod.Put => System.Net.Http.HttpMethod.Put,
        RustHttpMethod.Delete => System.Net.Http.HttpMethod.Delete,
        _ => System.Net.Http.HttpMethod.Post
    };

    /// <summary>
    /// Flatten an <see cref="HttpResponseMessage"/>'s response AND content headers
    /// into the binding <see cref="Header"/> list. Response headers must pass
    /// through — HW Cloud / routed / Gemini read response headers (X-Goog-Upload-*,
    /// Retry-After, credit balances). The core's lookup is case-insensitive.
    /// </summary>
    private static List<Header> CaptureHeaders(HttpResponseMessage response)
    {
        var headers = new List<Header>();
        foreach (var kvp in response.Headers)
        {
            headers.Add(new Header(@name: kvp.Key, @value: string.Join(",", kvp.Value)));
        }
        if (response.Content != null)
        {
            foreach (var kvp in response.Content.Headers)
            {
                headers.Add(new Header(@name: kvp.Key, @value: string.Join(",", kvp.Value)));
            }
        }
        return headers;
    }
}
